README.md
# mpy-utils v0.1

Utility programs for Micropython ...

* bin/mpy-upload : copy files in to a device using only the REPL.
* bin/mpy-sync : copy whole directory structure 
* bin/mpy-fuse : mount a device into the filesystem using FUSE ...

This package is also available through PyPI, see https://pypi.python.org/pypi/mpy-utils/


debian/changelog
mpy-utils (0.1.5) unstable; urgency=low

  * Initial Release.

 -- Paul Dwerryhouse <paul@dwerryhouse.com.au>  Wed, 26 Apr 2017 21:12:14 +1000


mpy-fuse.py
#!/usr/bin/env python3

import os
import sys
import errno
import argparse
import serial
import stat
import posix
import fuse

parser = argparse.ArgumentParser(
    description="FUSE mount a device using only the REPL"
)
parser.add_argument('--port', default='/dev/ttyUSB0', help='serial port device')
parser.add_argument('--baud', default=115200, type=int, help='port speed in baud')
parser.add_argument('--delay', default=100.0, type=float, help='delay between lines (ms)')
parser.add_argument('mount_point')
args = parser.parse_args()

port = serial.Serial(args.port, args.baud, timeout=2)

BUFSIZ = 50

def remote_initialize():
    port.write(b"\x04");
    while (port.read(100)): pass
    port.write(b"import os, sys\r")
    s = port.read(100)
    if not s.endswith(b"\r\n>>> "): raise("wot")

def safe_eval(s):
    try:
        return eval(s, {"__builtins__": {}}, {})
    except SyntaxError as e:
        print("Got %s" % e)

def remote_command(command):
    port.write(b"try: print(repr(%s))\rexcept Exception as e: print(b'!'+repr(e))\r\r" % command)
    port.readline()
    port.readline()
    port.readline()
    retn = port.readline()
    if retn.startswith(b'!'):
        return retn[1:]
    else:
        return safe_eval(retn)

def remote_function(func, *args):
    command = func + repr(tuple(args)).encode("ASCII")
    return remote_command(command)

def remote_filehandle(fh, func, *args):
    command = b"_fh%d=" % fh + func + repr(tuple(args)).encode("ASCII")
    port.write(command + b"\r")
    port.readline()

class MicroPythonOps(fuse.Operations):
    def __init__(self):
        remote_initialize()
        print("READY")
        self.fh_count = 0

    # Filesystem methods
    # ==================

    def access(self, path, mode):
        return 0

    def chmod(self, path, mode):
        raise fuse.FuseOSError(errno.EACCES)

    def chown(self, path, uid, gid):
        raise fuse.FuseOSError(errno.EACCES)

    def getattr(self, path, fh=None):
        s = remote_function(b'os.stat', path)
        if type(s) is bytes:
           raise fuse.FuseOSError(errno.ENOENT)

        mode = (stat.S_IFDIR | 0o755) if stat.S_ISDIR(s[0]) else (stat.S_IFREG | 0o644);
        return {
          'st_mode': mode,
          'st_nlink': 2,
          'st_size': s[6] or 4096,
          'st_uid': os.getuid(),
          'st_gid': os.getgid()
        }

    def readdir(self, path, fh):
        dirents = ['.', '..']
        dirents.extend(remote_function(b'os.listdir', path))
        return dirents

    def readlink(self, path):
        return path

    def rmdir(self, path):
        return remote_function(b'os.rmdir', path)

    def mkdir(self, path, mode):
        return remote_function(b'os.mkdir', path)

    def statfs(self, path):
        r = remote_function(b'os.statvfs', path)
        return dict(zip(['f_bsize','f_frsize','f_blocks','f_bfree','f_bavail','f_files','f_ffree','f_favail','f_fsid','f_flag'], r))

    def unlink(self, path):
        return remote_function(b'os.remove', path)

    def rename(self, old, new):
        return remote_function(b'os.rename', old, new)

    # File methods
    # ============

    def open(self, path, flags):
        print("open %s %s" % (path, flags))
        self.fh_count += 1
        mode = "rb" if (flags & posix.O_RDONLY) else "rb+"
        remote_filehandle(self.fh_count, b'open', path, mode)
        return self.fh_count

    def create(self, path, mode, fi=None):
        print("create %s %s" % (path, mode))
        self.fh_count += 1
        remote_filehandle(self.fh_count, b'open', path, "wb")
        return self.fh_count

    def read(self, path, length, offset, fh):
        remote_function(b'_fh%d.seek' % fh, offset)
        buf = b''
        while len(buf) < length:
            size = length - len(buf) if length - len(buf) < BUFSIZ else BUFSIZ;
            r = remote_function(b'_fh%d.read' % fh, size)
            if r is None or len(r) == 0: break
            buf += r
            print(len(buf))
        return buf

    def write(self, path, buf, offset, fh):
        length = len(buf)
        remote_function(b'_fh%d.seek' % fh, offset)
        total = 0
        while total < length:
            size = length - total if length - total < BUFSIZ else BUFSIZ;
            total += remote_function(b'_fh%d.write' % fh, buf[total:total+size])
        return total

    def flush(self, path, fh):
        return remote_command(b'_fh%d.flush()' % fh)

    def release(self, path, fh):
        return remote_command(b'_fh%d.close()' % fh)

    def fsync(self, path, fdatasync, fh):
        raise fuse.FuseOSError(errno.EACCES)
        return self.flush(path, fh)


def main(mount_point):
    fuse.FUSE(
        operations=MicroPythonOps(),
        fsname="mpy-fuse:%s" % args.port,
        mountpoint=mount_point,
        nothreads=True,
        foreground=True
    )

if __name__ == '__main__':
    main(args.mount_point)


mpy_utils/replcontrol.py
import serial
import string
import atexit
import time

class ReplControl(object):

    def __init__(self, port='/dev/ttyUSB0', baud=115200, delay=0, debug=False, reset=True):
        self.port = serial.Serial(port, baud, timeout=2)
        self.buffer = b""
        self.delay = delay
        self.debug = debug
        self.initialize()

        if reset:
            atexit.register(self.reset)

    def response(self, end=b"\x04"):
        while True:
            bytes_to_read = self.port.inWaiting()
            if not bytes_to_read:
                time.sleep(self.delay / 1000.0)

            self.buffer += self.port.read(bytes_to_read)
            try:
                r, self.buffer = self.buffer.split(end, 1)
                return r
            except ValueError:
                pass

    def initialize(self):
        # break, break, raw mode, reboot
        self.port.write(b"\x03\x03\x01\x04")
        start = time.time()
        while True:
            resp = self.port.read_all()
            if resp.endswith(b"\r\n>"):
                break
            elif time.time() - start > 3:
                if self.debug:
                    print("Forcefully breaking the boot.py")
                self.port.write(b"\x03\x03")
            time.sleep(self.delay / 1000.0)
        self.port.flushInput()

    def reset(self):
        self.port.write(b"\x02\x03\x03\x04")

    def command(self, cmd):
        if self.debug:
            print(">>> %s" % cmd)
        self.port.write(cmd.encode("ASCII") + b"\x04")
        time.sleep(self.delay / 1000.0)
        ret = self.response()
        err = self.response(b"\x04>")

        if ret.startswith(b'OK'):
            if err:
                if self.debug: print("<<< %s" % err)
                return err
            elif len(ret) > 2:
                if self.debug: print("<<< %s" % ret[2:])
                try:
                    return eval(ret[2:], {"__builtins__": {}}, {})
                except SyntaxError as e:
                    return e
            else:
                return None

    def statement(self, func, *args):
        return self.command(func + repr(tuple(args)))

    def function(self, func, *args):
        command = "print(repr(%s))" % (func + repr(tuple(args)))
        return self.command(command)

    def variable(self, func, *args):
        return ReplControlVariable(self, func, *args)


class ReplControlVariable(object):

    names = [ '_%s%s' % (x,y) for x in string.ascii_lowercase for y in string.ascii_lowercase ]

    def __init__(self, control, func, *args):
        self.control = control
        self.name = self.__class__.names.pop(0)
        self.control.statement("%s=%s" % (self.name, func), *args)

    def get_name(self):
        return self.name

    def method(self, method, *args):
        return self.control.function("%s.%s" % (self.name, method), *args)

    def __del__(self):
        self.control.command("del %s" % self.name)
        self.__class__.names.append(self.name)


mpy_utils/replfuseops.py
import errno
import stat
import posix
import fuse
import os

BUFSIZ = 50

def join_path(base, path):
    if path == '/': return base
    else: return base + path

class ReplFuseOps(fuse.Operations):
    def __init__(self, remote, base_path=''):
        self.remote = remote
        self.base_path = base_path
        self.filehandles = []
        self.remote.command("import os")

    # Filesystem methods
    # ==================

    def getattr(self, path, fh=None):
        s = self.remote.function('os.stat', join_path(self.base_path, path))
        if type(s) is bytes:
           raise fuse.FuseOSError(errno.ENOENT)

        mode = (stat.S_IFDIR | 0o750) if stat.S_ISDIR(s[0]) else (stat.S_IFREG | 0o640);
        return {
          'st_mode': mode,
          'st_nlink': 2,
          'st_size': s[6],
          'st_uid': os.getuid(),
          'st_gid': os.getgid()
        }

    def readdir(self, path, fh):
        dirents = ['.', '..']
        dirents.extend(self.remote.function('os.listdir', join_path(self.base_path, path)))
        return dirents

    def readlink(self, path):
        return path

    def rmdir(self, path):
        return self.remote.function('os.rmdir', join_path(self.base_path, path))

    def mkdir(self, path, mode):
        return self.remote.function('os.mkdir', join_path(self.base_path, path))

    def statfs(self, path):
        r = self.remote.function('os.statvfs', join_path(self.base_path, path))
        return {
            "f_bsize": r[0],
            "f_frsize": r[1],
            "f_blocks": r[2],
            "f_bfree": r[3],
            "f_bavail": r[4],
            "f_flag": 0
        }

    def unlink(self, path):
        return self.remote.function('os.remove', join_path(self.base_path, path))

    def rename(self, old, new):
        return self.remote.function('os.rename', join_path(self.base_path, old), join_path(self.base_path, new))

    # File methods
    # ============

    def _open(self, path, mode):
        num = len(self.filehandles)
        var = self.remote.variable('open', join_path(self.base_path, path), mode)
        self.filehandles.append(( var, path ))
        return num

    def open(self, path, flags):
        if flags & posix.O_WRONLY: mode = "wb"
        elif flags & posix.O_RDWR: mode = "rb+"
        else: mode = "rb"
        return self._open(path, mode)

    def create(self, path, mode, fi=None):
        return self._open(path, "wb")

    def read(self, path, length, offset, fh):
        self.filehandles[fh][0].method('seek', offset)
        buf = b''
        while len(buf) < length:
            size = length - len(buf) if length - len(buf) < BUFSIZ else BUFSIZ;
            r = self.filehandles[fh][0].method('read', int(size))
            if r is None or len(r) == 0: break
            buf += r
        return buf

    def write(self, path, buf, offset, fh):
        length = len(buf)
        self.filehandles[fh][0].method('seek', offset)
        total = 0
        while total < length:
            size = length - total if length - total < BUFSIZ else BUFSIZ;
            total += self.filehandles[fh][0].method('write', buf[total:total+size])
        return total

    def truncate(self, path, size):
        if size == 0:
            fh = self._open(path, "wb")
            self.filehandles[fh][0].method('close')
            self.filehandles[fh] = None
        else:
            # not implemented
            raise fuse.FuseOSError(errno.ENOTSUP)

    def flush(self, path, fh):
        return self.filehandles[fh][0].method('flush')

    def release(self, path, fh):
        self.filehandles[fh][0].method('close')
        self.filehandles[fh] = None

    def fsync(self, path, fdatasync, fh):
        for fh in self.filehandles:
            if fh is not None and fh[1] == path:
                fh[0].method('flush')


upload.py
#!/usr/bin/env python

import serial
import time
import argparse
import os.path

parser = argparse.ArgumentParser(
    description="upload files to a device using only the REPL"
)
parser.add_argument('--port', default='/dev/ttyUSB0', help='serial port device')
parser.add_argument('--baud', default=115200, type=int, help='port speed in baud')
parser.add_argument('--delay', default=100.0, type=float, help='delay between lines (ms)')
parser.add_argument('--interrupt', action='store_true', help='send soft interrupt (control-C) before upload')
parser.add_argument('--reset', action='store_true', help='send soft reset (control-D) after upload')
parser.add_argument('files', nargs='*', type=argparse.FileType('rb'))
args = parser.parse_args()

port = serial.Serial(args.port, args.baud)

if args.interrupt:
  port.write('\x03')

for fh in args.files:
  path, name = os.path.split(fh.name)
  port.write('_fh = open(%s, "w")\r' % repr(name))

  while True:
    s = fh.read(50)
    if len(s) == 0: break
    port.write("_fh.write(%s)\r" % repr(s))
    time.sleep(args.delay/1000.0)

  port.write('_fh.close()\r')
  fh.close()

if args.reset:
  port.write('\x04')

port.close()
